core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -81,9 +86,17 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
       val responseQueue = new ArrayBlockingQueue[Req#Response](requests.size)
 
       requests.foreach { request =>
-        send(request) { response =>
-          responseQueue.add(response)
-          countDownLatch.countDown()
+        inFlightRequests.acquire()
+        try {
+          send(request) { response =>
+            responseQueue.add(response)
+            inFlightRequests.release()
```

Comment:
We should probably add a comment that no method that can throw an exception can be called after `inFlightRequests.release()` or else we will release twice.

Reply:
A boolean variable can be introduced which records whether inFlightRequests has been released.
This boolean is checked in the finally block.

Reply:
@tedyu it seems like the current PR solution is simpler.

Reply:
Using the boolean, even if any call, following inFlightRequests.release(), throws exception, there wouldn't be double release.

Reply:
@tedyu the short answer is I don't believe your flag approach works. What's proposed in the PR is correct.

Something that's easy to forget is that the following block is processed asynchronously from the `org.apache.zookeeper.EventThread`:
```
{ response =>
  responseQueue.add(response)
  inFlightRequests.release()
  countDownLatch.countDown()
}
```
On the other hand, the rest of the try/catch block in `handleRequests`, or your proposed try/finally block, is processed by the calling thread.

I think what you're suggesting is:
```
flag = false
requests.foreach { request =>
  inFlightRequests.acquire()
  try {
    send(request) { response =>
      responseQueue.add(response)
      inFlightRequests.release()
      countDownLatch.countDown()
      flag = true
    }
  } finally {
    if (!flag)
      inFlightRequests.release()
  }
}
```
In this snippet, the calling thread will quickly run through the whole try block quickly since it's asynchronous, execute the finally block, and check the flag you're proposing. By this time, the flag may or may not be set depending on how quickly zookeeper processes our request and our asynchronous block gets processed.

The behavior we actually want is split into two scenarios:
1. The happy case: the underlying `org.apache.zookeeper.ZooKeeper` client is able to successfully send out a request asynchronously to ZooKeeper and the `org.apache.zookeeper.EventThread` processes the asynchronous block after retrieving a response. We should only update the concurrency primitives after retrieving the response.
2. The sad case: the underlying `org.apache.zookeeper.ZooKeeper` client immediately throws an exception from the calling thread before even sending out a request to ZooKeeper, in which case we should only call `inFlightRequests.release`.

Reply:
Makes sense.

Reply:
@ijuma regarding https://github.com/apache/kafka/pull/3860#discussion_r148135333

I don't think a double-release is possible. We either:
1. execute the block from the `org.apache.zookeeper.EventThread`, potentially resulting in an exception which the `org.apache.zookeeper.EventThread` would catch.
2. execute the catch block when the calling thread attempts to `send`.

The only risk I see is if an exception somehow gets thrown from the asynchronous block before the call to `inFlightRequests.release`, not after. In this scenario, the semaphore would not get decremented for the problematic response callback.

Reply:
Thanks for the explanation. So the release in the callback should be in a try/finally?